README.rst
=================
django-audiofield
=================

Django-audiofield is a Django application which allows audio file upload and conversion to mp3, wav and ogg format.
It also makes it easy to play the audio files into your django application, for this we integrated a HTML5 and Flash audio player 'SoundManager2'

The goal of this project is to quickly manage audio files into your django project and make it easy for admins and users to listen to them.


.. image:: https://github.com/Star2Billing/django-audiofield/raw/master/docs/source/_static/django-admin-audiofield.png

.. image:: https://github.com/Star2Billing/django-audiofield/raw/master/docs/source/_static/django-admin-audiofield-upload.png

More information about Soundmanager2 : http://www.schillmania.com/projects/soundmanager2/


Installation
============

Install Django-Audiofield::

    python setup.py install


Dependencies
------------

Install dependencies on Debian::

    apt-get -y install libsox-fmt-mp3 libsox-fmt-all mpg321 dir2ogg libav-tools


Install dependencies on Redhat/CentOS::

    yum -y install python-setuptools libsox-fmt-mp3 libsox-fmt-all mpg321 dir2ogg


Install avconv on Redhat/CentOS::

    git clone git://git.libav.org/libav.git
    cd libav
    sudo ./configure --disable-yasm
    sudo make
    sudo make install


Settings
========

in your settings.py file::

    # Set Following variable
    #MEDIA_ROOT = ''
    #MEDIA_URL = ''

    In MIDDLEWARE_CLASSES add 'audiofield.middleware.threadlocals.ThreadLocals'

    In INSTALLED_APPS add 'audiofield'

    # Frontend widget values
    CHANNEL_TYPE_VALUE = 0  # 0-Keep original, 1-Mono, 2-Stereo

    FREQ_TYPE_VALUE = 8000  # 0-Keep original, 8000-8000Hz, 16000-16000Hz, 22050-22050Hz,
                         # 44100-44100Hz, 48000-48000Hz, 96000-96000Hz

    CONVERT_TYPE_VALUE = 0 # 0-Keep original, 1-Convert to MP3, 2-Convert to WAV, 3-Convert to OGG


Usage
=====

Add the following lines in your models.py file::

    from django.conf import settings
    from audiofield.fields import AudioField
    import os.path

    # Add the audio field to your model
    audio_file = AudioField(upload_to='your/upload/dir', blank=True,
                            ext_whitelist=(".mp3", ".wav", ".ogg"),
                            help_text=("Allowed type - .mp3, .wav, .ogg"))

    # Add this method to your model
    def audio_file_player(self):
        """audio player tag for admin"""
        if self.audio_file:
            file_url = settings.MEDIA_URL + str(self.audio_file)
            player_string = '<ul class="playlist"><li style="width:250px;">\
            <a href="%s">%s</a></li></ul>' % (file_url, os.path.basename(self.audio_file.name))
            return player_string
    audio_file_player.allow_tags = True
    audio_file_player.short_description = _('Audio file player')


Add the following lines in your admin.py::


    from your_app.models import your_model_name

    # add 'audio_file_player' tag to your admin view
    list_display = (..., 'audio_file_player', ...)
    actions = ['custom_delete_selected']

    def custom_delete_selected(self, request, queryset):
        #custom delete code
        n = queryset.count()
        for i in queryset:
            if i.audio_file:
                if os.path.exists(i.audio_file.path):
                    os.remove(i.audio_file.path)
            i.delete()
        self.message_user(request, _("Successfully deleted %d audio files.") % n)
    custom_delete_selected.short_description = "Delete selected items"

    def get_actions(self, request):
        actions = super(AudioFileAdmin, self).get_actions(request)
        del actions['delete_selected']
        return actions


If you are not using the installation script, please copy following template
file to your template directory::

    cp audiofield/templates/common_audiofield.html /path/to/your/templates/directory/


Add the following in your template files (like admin/change_form.html, admin/change_list.html etc.
in which you are using audio field type)::


    {% block extrahead %}
    {{ block.super }}
        {% include "common_audiofield.html" %}
    {% endblock %}


Then perform following commands to create the table and collect the static files::

    ./manage.py syncdb
    ./manage.py collectstatic


Create audiofield.log file::

    touch /var/log/audio-field.log


Contributing
============

If you've found a bug, implemented a feature or customized the template and
think it is useful then please consider contributing. Patches, pull requests or
just suggestions are welcome!

Source code: http://github.com/Star2Billing/django-audiofield

Bug tracker: https://github.com/Star2Billing/django-audiofield/issues


Documentation
=============

Documentation is available on 'Read the Docs':
http://django-audiofield.readthedocs.org


Credit
======

Django-audiofield is a Star2Billing-Sponsored Community Project, for more information visit
http://www.star2billing.com or email us at info@star2billing.com


License
=======

django-audiofield is licensed under MIT, see `MIT-LICENSE.txt`.


CHANGES.txt
Changelog
=========


0.5.8 (2013-10-26)
------------------

* Fix convertions to OGG
* Update install instructions


0.5.7 (2013-10-25)
------------------

* Fix convertion from MP3 to WAV and from OGG to WAV


0.5.6 (2013-10-24)
------------------

* Change permission name to view_audiofile


0.5.5 (2013-03-07)
------------------

* Update trans string


0.5.4 (2012-12-05)
------------------

* Fix sox conversion


0.5.3 (2012-12-04)
------------------

* Replace commands by subprocess


0.5.2 (2012-10-16)
------------------

* Add celery requirements


0.5 (2012-10-04)
------------------

* Fix requirements


0.4 (2012-10-03)
------------------

* Improve setup.py and update manifest
* Update README.rst
* Fix PEP8


0.3 (2011-12-09)
----------------

* Import project


audiofield/settings.py
import os


# Django settings for project.
DEBUG = False
TEMPLATE_DEBUG = False

ADMINS = (
     ('Your Name', 'your_email@domain.com'),
)
MANAGERS = ADMINS

SERVER_EMAIL = 'audiofield@localhost.com'

APPLICATION_DIR = os.path.dirname(globals()['__file__'])

DATABASES = {
    'default': {
        # 'postgresql_psycopg2','postgresql','mysql','sqlite3','oracle'
        'ENGINE': 'django.db.backends.sqlite3',
        # Database name or path to database file if using sqlite3.
        'NAME': APPLICATION_DIR + '/database/newfies-dialer.db',
        'USER': '',                      # Not used with sqlite3.
        'PASSWORD': '',                  # Not used with sqlite3.
        'HOST': '',                      # Not used with sqlite3.
        'PORT': '',                      # Not used with sqlite3.
        'OPTIONS': {
           'init_command': 'SET storage_engine=INNODB',
        }
    }
}

CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
        'LOCATION': '/var/tmp/django_cache',
    }
}

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# On Unix systems, a value of None will cause Django to use the same
# timezone as the operating system.
# If running in a Windows environment this must be set to the same as your
# system time zone.
TIME_ZONE = 'America/Chicago'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale
USE_L10N = True



# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/home/media/media.lawrence.com/static/"
STATIC_ROOT = os.path.join(os.path.dirname(__file__), 'static')

# URL prefix for static files.
# Example: "http://media.lawrence.com/static/"
STATIC_URL = '/static/'

# Absolute path to the directory that holds media.
# Example: "/home/media/media.lawrence.com/"
MEDIA_ROOT = os.path.join(os.path.dirname(__file__), 'usermedia')

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash if there is a path component (optional in other cases).
# Examples: "http://media.lawrence.com", "http://example.com/media/"
MEDIA_URL = '/mediafiles/'

# URL prefix for admin media -- CSS, JavaScript and images. Make sure to use a
# trailing slash.
# Examples: "http://foo.com/media/", "/media/".
ADMIN_MEDIA_PREFIX = STATIC_URL + 'admin/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    # os.path.join(APPLICATION_DIR, "resources"),
    ("audiofield", os.path.join(APPLICATION_DIR, "resources")),
)



# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
#    'django.contrib.staticfiles.finders.DefaultStorageFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = 'ujau$^uei_ak=@-v8va(&@q_sc0^1nn*qpwyc-776n&qoam@+v'

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
    'django.template.loaders.eggs.Loader',
)

MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'common.filter_persist_middleware.FilterPersistMiddleware',
    'audiofield.middleware.threadlocals.ThreadLocals',
)


TEMPLATE_CONTEXT_PROCESSORS = (
    "django.contrib.auth.context_processors.auth",
    "django.core.context_processors.debug",
    "django.core.context_processors.i18n",
    "django.core.context_processors.media",
    "django.core.context_processors.static",
    "django.core.context_processors.csrf",
    "django.contrib.messages.context_processors.messages",
)

ROOT_URLCONF = 'urls'

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates" or
    # "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    os.path.join(APPLICATION_DIR, 'templates'),
)

INTERNAL_IPS = ('127.0.0.1',)



INSTALLED_APPS = (
    #admin tool apps
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    # Uncomment the next line to enable the admin:
    'django.contrib.admin',
    'common',
    'audiofield',
)



#LANGUAGES
#===========
gettext = lambda s: s
LANGUAGES = (
    ('en', gettext('English')),
    #('fr', gettext('French')),
    ('es', gettext('Spanish')),
    ('pt', gettext('Portuguese')),
)



# Frontend widget values
CHANNEL_TYPE_VALUE = 1  # 0-Keep original, 1-Mono, 2-Stereo

FREQ_TYPE_VALUE = 8000  # 0-Keep original, 8000-8000Hz, 16000-16000Hz, 22050-22050Hz,
                     # 44100-44100Hz, 48000-48000Hz, 96000-96000Hz

CONVERT_TYPE_VALUE = 2 # 0-Keep original, 1-Convert to MP3, 2-Convert to WAV, 3-Convert to OGG

AUDIO_DEBUG = False

#IMPORT LOCAL SETTINGS
#=====================
#from settings_local import *

